<a href="https://colab.research.google.com/github/FatimaaShahid/Machine-Learning-Labs-CS-324-/blob/main/Lab2/ML_LAB_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
from pandas import Series, DataFrame

In [3]:
s = pd.Series([12,23,44,65,78])
s

0    12
1    23
2    44
3    65
4    78
dtype: int64

In [4]:
s.index=[2,3,6,7,8]

In [ ]:
data = {
    'names' = ['ali','Sara','asra'],
    'ages'= [22,21,23],
    'cgpa'=[2.1,3,3.5]}

In [ ]:
df = pd.DataFrame(data)
df.info()

In [5]:
from google.colab import files
uploaded = files.upload()


Saving task-1.csv to task-1 (1).csv


In [7]:
df = pd.read_csv('task-1.csv')
df.head()

order_id  order_date       product     category   price  quantity  \
0      1001  2024-01-05        Laptop          NaN  120000         1   
1      1002  2024-01-06  Mobile Phone  Electronics   60000         2   
2      1003  2024-01-07    Headphones  Accessories    5000         3   
3      1004  2024-01-10  Office Chair    Furniture   25000         1   
4      1005  2024-02-02        Laptop  Electronics  120000         1   

        city    payment_method  
0    Karachi       Credit Card  
1     Lahore  Cash on Delivery  
2  Islamabad        Debit Card  
3        khi       Credit Card  
4     Lahore       Credit Card

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        11 non-null     int64 
 1   order_date      11 non-null     object
 2   product         11 non-null     object
 3   category        10 non-null     object
 4   price           11 non-null     int64 
 5   quantity        11 non-null     int64 
 6   city            11 non-null     object
 7   payment_method  11 non-null     object
dtypes: int64(3), object(5)
memory usage: 836.0+ bytes


In [9]:
df.tail()

order_id  order_date       product     category  price  quantity  \
6       1007  2024-02-15  Mobile Phone  Electronics  60000         1   
7       1008  2024-03-01       Monitor  Electronics  30000         2   
8       1009  2024-03-05    Headphones  Accessories   5000         4   
9       1010  2024-03-10     Desk Lamp  Accessories   3000         2   
10      1010  2024-03-10     Desk Lamp  Accessories   3000         2   

         city    payment_method  
6     Karachi        Debit Card  
7      Lahore       Credit Card  
8     KARACHI  Cash on Delivery  
9   Islamabad        Debit Card  
10  Islamabad        Debit Card

In [10]:
df.tail(6)

order_id  order_date       product     category  price  quantity  \
5       1006  2024-02-05          Desk    Furniture  40000         1   
6       1007  2024-02-15  Mobile Phone  Electronics  60000         1   
7       1008  2024-03-01       Monitor  Electronics  30000         2   
8       1009  2024-03-05    Headphones  Accessories   5000         4   
9       1010  2024-03-10     Desk Lamp  Accessories   3000         2   
10      1010  2024-03-10     Desk Lamp  Accessories   3000         2   

         city    payment_method  
5   Islamabad  Cash on Delivery  
6     Karachi        Debit Card  
7      Lahore       Credit Card  
8     KARACHI  Cash on Delivery  
9   Islamabad        Debit Card  
10  Islamabad        Debit Card

In [11]:
df.shape

(11, 8)

In [12]:
df['category'].unique()

array([nan, 'Electronics', 'Accessories', 'Furniture'], dtype=object)

In [14]:
df.dtypes

order_id           int64
order_date        object
product           object
category          object
price              int64
quantity           int64
city              object
payment_method    object
dtype: object

In [15]:
df.describe()

order_id          price   quantity
count    11.000000      11.000000  11.000000
mean   1005.909091   42818.181818   1.818182
std       3.176619   43618.386448   0.981650
min    1001.000000    3000.000000   1.000000
25%    1003.500000    5000.000000   1.000000
50%    1006.000000   30000.000000   2.000000
75%    1008.500000   60000.000000   2.000000
max    1010.000000  120000.000000   4.000000

In [16]:
df.isna().sum() #number of missing val in a col

order_id          0
order_date        0
product           0
category          1
price             0
quantity          0
city              0
payment_method    0
dtype: int64

In [17]:
df[df['category'].isna()]

order_id  order_date product category   price  quantity     city  \
0      1001  2024-01-05  Laptop      NaN  120000         1  Karachi   

  payment_method  
0    Credit Card

In [20]:
df.loc[0,'category'] = 'Electronics' #[val,col]
df[df['category'].isna()]

In [19]:
df[df['category'].isna()]

Empty DataFrame
Columns: [order_id, order_date, product, category, price, quantity, city, payment_method]
Index: []

In [ ]:
#or we can drop that row using
#df.dropna()
#df['payment_method].fillna('Unknown',inplace = True) we can replace the values

In [21]:
df.duplicated().sum()

np.int64(1)

In [22]:
df = df.drop_duplicates()

In [23]:
df.duplicated().sum()

np.int64(0)

In [24]:
df.tail()

order_id  order_date       product     category  price  quantity  \
5      1006  2024-02-05          Desk    Furniture  40000         1   
6      1007  2024-02-15  Mobile Phone  Electronics  60000         1   
7      1008  2024-03-01       Monitor  Electronics  30000         2   
8      1009  2024-03-05    Headphones  Accessories   5000         4   
9      1010  2024-03-10     Desk Lamp  Accessories   3000         2   

        city    payment_method  
5  Islamabad  Cash on Delivery  
6    Karachi        Debit Card  
7     Lahore       Credit Card  
8    KARACHI  Cash on Delivery  
9  Islamabad        Debit Card

In [ ]:
#we have inconsitency in KARACHI and karachi

In [25]:
df['city'] = df['city'].str.lower().str.strip()

/tmp/ipython-input-4153634180.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['city'] = df['city'].str.lower().str.strip()


In [26]:
df.tail()

order_id  order_date       product     category  price  quantity  \
5      1006  2024-02-05          Desk    Furniture  40000         1   
6      1007  2024-02-15  Mobile Phone  Electronics  60000         1   
7      1008  2024-03-01       Monitor  Electronics  30000         2   
8      1009  2024-03-05    Headphones  Accessories   5000         4   
9      1010  2024-03-10     Desk Lamp  Accessories   3000         2   

        city    payment_method  
5  islamabad  Cash on Delivery  
6    karachi        Debit Card  
7     lahore       Credit Card  
8    karachi  Cash on Delivery  
9  islamabad        Debit Card

In [27]:
df.head(20)

order_id  order_date       product     category   price  quantity  \
0      1001  2024-01-05        Laptop  Electronics  120000         1   
1      1002  2024-01-06  Mobile Phone  Electronics   60000         2   
2      1003  2024-01-07    Headphones  Accessories    5000         3   
3      1004  2024-01-10  Office Chair    Furniture   25000         1   
4      1005  2024-02-02        Laptop  Electronics  120000         1   
5      1006  2024-02-05          Desk    Furniture   40000         1   
6      1007  2024-02-15  Mobile Phone  Electronics   60000         1   
7      1008  2024-03-01       Monitor  Electronics   30000         2   
8      1009  2024-03-05    Headphones  Accessories    5000         4   
9      1010  2024-03-10     Desk Lamp  Accessories    3000         2   

        city    payment_method  
0    karachi       Credit Card  
1     lahore  Cash on Delivery  
2  islamabad        Debit Card  
3        khi       Credit Card  
4     lahore       Credit Card  
5  islamabad  Cash on Delivery  
6    karachi        Debit Card  
7     lahore       Credit Card  
8    karachi  Cash on Delivery  
9  islamabad        Debit Card

In [28]:
df['city'] = df['city'].replace({'khi':'karachi'})

/tmp/ipython-input-1491663184.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['city'] = df['city'].replace({'khi':'karachi'})


In [29]:
df['order_date'] = pd.to_datetime(df['order_date'])

/tmp/ipython-input-2765250300.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['order_date'] = pd.to_datetime(df['order_date'])


In [31]:
df[df['price']>300000]

Empty DataFrame
Columns: [order_id, order_date, product, category, price, quantity, city, payment_method]
Index: []

In [32]:
df['total_sales'] = df['price']*df['quantity'] #creating new feature

/tmp/ipython-input-492888806.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['total_sales'] = df['price']*df['quantity'] #creating new feature


In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, 0 to 9
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        10 non-null     int64         
 1   order_date      10 non-null     datetime64[ns]
 2   product         10 non-null     object        
 3   category        10 non-null     object        
 4   price           10 non-null     int64         
 5   quantity        10 non-null     int64         
 6   city            10 non-null     object        
 7   payment_method  10 non-null     object        
 8   total_sales     10 non-null     int64         
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 1.1+ KB


In [39]:
df['sales_category'] = pd.cut(df['total_sales'],
                              bins =[0,20000,50000,100000,float('inf')],
                              labels =['Low','Medium','High','Very High'])

/tmp/ipython-input-2217274157.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sales_category'] = pd.cut(df['total_sales'],


In [40]:
df['discount'] = df['category'].map({'Electronics':0.05,
                                     'Furniture':0.10,
                                     'Accessories':0.15})

/tmp/ipython-input-733248297.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['discount'] = df['category'].map({'Electronics':0.05,


In [41]:
df['final_sales'] = df['total_sales']*(1-df['discount'])

/tmp/ipython-input-2383935347.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['final_sales'] = df['total_sales']*(1-df['discount'])


In [42]:
df['month'] = df['order_date'].dt.month

/tmp/ipython-input-1924987876.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['month'] = df['order_date'].dt.month


In [43]:
df['weekday'] = df['order_date'].dt.day_name()

/tmp/ipython-input-4101422905.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['weekday'] = df['order_date'].dt.day_name()


In [44]:
df.head(20)

order_id order_date       product     category   price  quantity  \
0      1001 2024-01-05        Laptop  Electronics  120000         1   
1      1002 2024-01-06  Mobile Phone  Electronics   60000         2   
2      1003 2024-01-07    Headphones  Accessories    5000         3   
3      1004 2024-01-10  Office Chair    Furniture   25000         1   
4      1005 2024-02-02        Laptop  Electronics  120000         1   
5      1006 2024-02-05          Desk    Furniture   40000         1   
6      1007 2024-02-15  Mobile Phone  Electronics   60000         1   
7      1008 2024-03-01       Monitor  Electronics   30000         2   
8      1009 2024-03-05    Headphones  Accessories    5000         4   
9      1010 2024-03-10     Desk Lamp  Accessories    3000         2   

        city    payment_method  total_sales sales_category  discount  \
0    karachi       Credit Card       120000      Very High      0.05   
1     lahore  Cash on Delivery       120000      Very High      0.05   
2  islamabad        Debit Card        15000            Low      0.15   
3    karachi       Credit Card        25000         Medium      0.10   
4     lahore       Credit Card       120000      Very High      0.05   
5  islamabad  Cash on Delivery        40000         Medium      0.10   
6    karachi        Debit Card        60000           High      0.05   
7     lahore       Credit Card        60000           High      0.05   
8    karachi  Cash on Delivery        20000            Low      0.15   
9  islamabad        Debit Card         6000            Low      0.15   

   final_sales  month    weekday  
0     114000.0      1     Friday  
1     114000.0      1   Saturday  
2      12750.0      1     Sunday  
3      22500.0      1  Wednesday  
4     114000.0      2     Friday  
5      36000.0      2     Monday  
6      57000.0      2   Thursday  
7      57000.0      3     Friday  
8      17000.0      3    Tuesday  
9       5100.0      3     Sunday

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, 0 to 9
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        10 non-null     int64         
 1   order_date      10 non-null     datetime64[ns]
 2   product         10 non-null     object        
 3   category        10 non-null     object        
 4   price           10 non-null     int64         
 5   quantity        10 non-null     int64         
 6   city            10 non-null     object        
 7   payment_method  10 non-null     object        
 8   total_sales     10 non-null     int64         
 9   sales_category  10 non-null     category      
 10  discount        10 non-null     float64       
 11  final_sales     10 non-null     float64       
 12  month           10 non-null     int32         
 13  weekday         10 non-null     object        
dtypes: category(1), datetime64[ns](1), float64(2), int32(1), int64(4),

In [46]:
df['final_sales'].sum()

np.float64(549350.0)

In [50]:
df.groupby('product')['final_sales'].sum().sort_values(ascending = False)

product
Laptop          228000.0
Mobile Phone    171000.0
Monitor          57000.0
Desk             36000.0
Headphones       29750.0
Office Chair     22500.0
Desk Lamp         5100.0
Name: final_sales, dtype: float64

In [52]:
df.groupby('category')['final_sales'].sum().sort_values(ascending = False)

category
Electronics    456000.0
Furniture       58500.0
Accessories     34850.0
Name: final_sales, dtype: float64

###Task 2

In [55]:
df2 = pd.read_csv('task-2.csv')
df2.head(15)

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    NaN        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0         NaN       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   11.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    NaN    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  
0    45.0          90         1  
1    40.0          85         1  
2    48.0          95         1  
3     NaN          80         1  
4    49.0         105         1  
5   -10.0          92         1  
6    44.0          88         1  
7    39.0          82         1  
8    47.0          90         1  
9    42.0          80         1  
10   45.0          90         1  
11   38.0          75         1  
12   50.0          98         1  
13   35.0          70         1  
14   40.0          85         1

In [54]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   student_id  15 non-null     int64  
 1   name        15 non-null     object 
 2   course      15 non-null     object 
 3   quiz1       14 non-null     float64
 4   quiz2       14 non-null     float64
 5   assignment  14 non-null     float64
 6   midterm     15 non-null     int64  
 7   final       14 non-null     float64
 8   attendance  15 non-null     int64  
 9   semester    15 non-null     int64  
dtypes: float64(4), int64(4), object(2)
memory usage: 1.3+ KB


In [58]:
df2[df2['final']<0]

student_id       name     course  quiz1  quiz2  assignment  midterm  final  \
5        4006  Hina Raza  Chemistry    8.0    8.0        17.0       27  -10.0   

   attendance  semester  
5          92         1

In [59]:
df2.loc[5,'final'] = 0

In [60]:
df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    NaN        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0         NaN       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   11.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    NaN    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  
0    45.0          90         1  
1    40.0          85         1  
2    48.0          95         1  
3     NaN          80         1  
4    49.0         105         1  
5     0.0          92         1  
6    44.0          88         1  
7    39.0          82         1  
8    47.0          90         1  
9    42.0          80         1  
10   45.0          90         1  
11   38.0          75         1  
12   50.0          98         1  
13   35.0          70         1  
14   40.0          85         1

In [64]:
df2.loc[3,'final'] = round(df2['final'].mean(),1)

In [65]:
df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    NaN        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0         NaN       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   11.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    NaN    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  
0    45.0          90         1  
1    40.0          85         1  
2    48.0          95         1  
3    40.1          80         1  
4    49.0         105         1  
5     0.0          92         1  
6    44.0          88         1  
7    39.0          82         1  
8    47.0          90         1  
9    42.0          80         1  
10   45.0          90         1  
11   38.0          75         1  
12   50.0          98         1  
13   35.0          70         1  
14   40.0          85         1

In [74]:
df2.loc[1,'quiz2'] = round(df2['quiz2'].mean(),1)

df2.loc[7,'quiz1'] = round(df2['quiz1'].mean(),1)
df2.loc[2,'assignment'] = round(df2['assignment'].mean(),1)
df2.loc[4,'quiz1'] = 10.0
df2.loc[4,'attendance'] = 100
df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    6.7        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0        15.7       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   10.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    7.5    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  total_score  
0    45.0          90         1        103.0  
1    40.0          85         1         88.7  
2    48.0          95         1        108.7  
3    40.1          80         1         87.1  
4    49.0         100         1        110.0  
5     0.0          92         1         60.0  
6    44.0          88         1         97.0  
7    39.0          82         1         88.6  
8    47.0          90         1        103.0  
9    42.0          80         1         91.0  
10   45.0          90         1        103.0  
11   38.0          75         1         82.0  
12   50.0          98         1        116.0  
13   35.0          70         1         76.0  
14   40.0          85         1         90.0

In [75]:
df2['total_score'] = df2['quiz1']+df2['quiz2']+df2['assignment']+df2['final']+df2['midterm']

df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    6.7        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0        15.7       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   10.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    7.5    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  total_score  
0    45.0          90         1        103.0  
1    40.0          85         1         88.7  
2    48.0          95         1        108.7  
3    40.1          80         1         87.1  
4    49.0         100         1        110.0  
5     0.0          92         1         60.0  
6    44.0          88         1         97.0  
7    39.0          82         1         88.5  
8    47.0          90         1        103.0  
9    42.0          80         1         91.0  
10   45.0          90         1        103.0  
11   38.0          75         1         82.0  
12   50.0          98         1        116.0  
13   35.0          70         1         76.0  
14   40.0          85         1         90.0

In [76]:
df2['grade'] = pd.cut(df2['total_score'],
                              bins =[0,20,40,60,80,100,120],
                              labels =['F','E','D','C','B','A'])
df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    6.7        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0        15.7       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   10.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    7.5    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  total_score grade  
0    45.0          90         1        103.0     A  
1    40.0          85         1         88.7     B  
2    48.0          95         1        108.7     A  
3    40.1          80         1         87.1     B  
4    49.0         100         1        110.0     A  
5     0.0          92         1         60.0     D  
6    44.0          88         1         97.0     B  
7    39.0          82         1         88.5     B  
8    47.0          90         1        103.0     A  
9    42.0          80         1         91.0     B  
10   45.0          90         1        103.0     A  
11   38.0          75         1         82.0     B  
12   50.0          98         1        116.0     A  
13   35.0          70         1         76.0     C  
14   40.0          85         1         90.0     B

In [78]:
import numpy as np

df2['result'] = np.where(df2['total_score'] > 60, 'Pass', 'Fail')
df2

student_id          name       course  quiz1  quiz2  assignment  midterm  \
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   
1         4002    Sara Ahmed      Physics    7.0    6.7        15.0       20   
2         4003    Omar Malik    Chemistry    9.0    8.0        15.7       28   
3         4004   Ayesha Khan  Mathematics    6.0    5.0        14.0       22   
4         4005   Usman Iqbal      Physics   10.0    7.0        18.0       26   
5         4006     Hina Raza    Chemistry    8.0    8.0        17.0       27   
6         4007      Zain Ali  Mathematics    7.0    7.0        16.0       23   
7         4008    Nadia Khan      Physics    7.5    6.0        15.0       21   
8         4009   Fahad Iqbal    Chemistry    8.0    7.0        16.0       25   
9         4010    Maryam Ali  Mathematics    7.0    6.0        14.0       22   
10        4011      Ali Khan  Mathematics    8.0    7.0        18.0       25   
11        4012   Hassan Raza      Physics    6.0    5.0        13.0       20   
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
13        4014     Omar Khan  Mathematics    5.0    6.0        12.0       18   
14        4015    Sara Ahmed      Physics    7.0    7.0        15.0       21   

    final  attendance  semester  total_score grade result  
0    45.0          90         1        103.0     A   Pass  
1    40.0          85         1         88.7     B   Pass  
2    48.0          95         1        108.7     A   Pass  
3    40.1          80         1         87.1     B   Pass  
4    49.0         100         1        110.0     A   Pass  
5     0.0          92         1         60.0     D   Fail  
6    44.0          88         1         97.0     B   Pass  
7    39.0          82         1         88.5     B   Pass  
8    47.0          90         1        103.0     A   Pass  
9    42.0          80         1         91.0     B   Pass  
10   45.0          90         1        103.0     A   Pass  
11   38.0          75         1         82.0     B   Pass  
12   50.0          98         1        116.0     A   Pass  
13   35.0          70         1         76.0     C   Pass  
14   40.0          85         1         90.0     B   Pass

In [82]:
df2.groupby('course')['total_score'].mean()


course
Chemistry      96.925
Mathematics    92.850
Physics        91.840
Name: total_score, dtype: float64

In [87]:
df2.groupby('grade')['total_score'].count()

/tmp/ipython-input-3258397987.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df2.groupby('grade')['total_score'].count()


grade
F    0
E    0
D    1
C    1
B    7
A    6
Name: total_score, dtype: int64

In [97]:
(df2['result'] == 'Fail').sum()


np.int64(1)

Insight:


###Average score per course

Chemistry	96.925

Mathematics	92.850

Physics	91.840

###Number of students per grade:
F:	0

E:	0

D:	1

C:	1

B:	7

A:	6

###sudents at risk
1 student have less than 60 marks

In [104]:
df2.loc[df2.groupby('course')['total_score'].idxmax()].sort_values(by = 'total_score',ascending = False)


student_id          name       course  quiz1  quiz2  assignment  midterm  \
12        4013  Fatima Malik    Chemistry    9.0    8.0        19.0       30   
4         4005   Usman Iqbal      Physics   10.0    7.0        18.0       26   
0         4001      Ali Khan  Mathematics    8.0    7.0        18.0       25   

    final  attendance  semester  total_score grade result  
12   50.0          98         1        116.0     A   Pass  
4    49.0         100         1        110.0     A   Pass  
0    45.0          90         1        103.0     A   Pass